### ISSUE TITLE
Support for Block Building

### ISSUE THREAD

**Issue by metachris** [paradigmxyz/reth#1923]
### Describe the feature

**It would be great if reth supports building blocks, and allows submitting to multiple mev-boost relays.**

Based on early reth metrics, it looks like it could be several times as performant as typical geth-based block builders, and perhaps provide some nice features on top!

At Flashbots we maintain a block builder based on geth, and would be happy to support the development of this functionality in reth with our experience and learnings.

### References

- [https://github.com/flashbots/builder](https://github.com/flashbots/builder) - geth-based block builder
- [https://flashbots.github.io/relay-specs](https://flashbots.github.io/relay-specs/) - APIs the relays implement for builders and the public
- [https://ethereum.github.io/builder-specs](https://ethereum.github.io/builder-specs/) - mev-boost core protocol APIs
- CL client SSE event to trigger block building
    - [https://github.com/ethereum/beacon-APIs/blob/master/apis/eventstream/index.yaml#L95-L124](https://github.com/ethereum/beacon-APIs/blob/master/apis/eventstream/index.yaml#L95-L124) - `payload_attributes` SSE event
    - [https://github.com/ethereum/execution-apis/blob/main/src/engine/shanghai.md#payloadattributesv2](https://github.com/ethereum/execution-apis/blob/main/src/engine/shanghai.md#payloadattributesv2) - `PayloadAttributesV2` structure
- Builder documentation: [Flashbots](https://docs.flashbots.net/flashbots-auction/searchers/advanced/rpc-endpoint), [builder0x69](https://docs.builder0x69.io/)

### Goals

1. Build blocks
    1. using transactions from mempool as well as directly sent transactions and bundles
    2. according proposer preferences: `gasLimit`, `feeRecipient`
    3. collect gas fees to the builder’s own `block.coinbase` address
    4. at end of the block, have a payment transaction from the builder (`block.coinbase`) to proposer feeRecipient
2. Submit the block to one or more mev-boost relays
3. Additional APIs to sending private transactions and bundles to the builder: `eth_callBundle`, `eth_sendBundle`, `eth_sendPrivateTransaction` ([[1]](https://docs.flashbots.net/flashbots-auction/searchers/advanced/rpc-endpoint#eth_sendprivatetransaction) [[2]](https://docs.alchemy.com/reference/eth-sendprivatetransaction) [[3]](https://www.quicknode.com/docs/ethereum/eth_sendPrivateTransaction)) / `eth_sendPrivateRawTransaction` ([[1]](https://docs.builder0x69.io/))

### How it works

1. Builder polls relay for the proposer registrations for the next epoch when block building is triggered 
    1. [https://flashbots.github.io/relay-specs/#/Builder/getValidators](https://flashbots.github.io/relay-specs/#/Builder/getValidators)
    2. [https://boost-relay.flashbots.net/relay/v1/builder/validators](https://boost-relay.flashbots.net/relay/v1/builder/validators)
    3. Note that multiple relays should be supported
2. The builder uses the `payload_attributes` SSE event subscription to a CL client to receive information about the current slot and the required payload attributes (timestamp, withdrawals, prevRandao) which are required for the produced block. 
    1. See also [PayloadAttributesV2 structure](https://github.com/ethereum/execution-apis/blob/main/src/engine/shanghai.md#payloadattributesv2) and the [SSE event docs](https://github.com/ethereum/beacon-APIs/blob/master/apis/eventstream/index.yaml#L95-L124)
3. Blocks are submitted to the relay(s) on an ongoing basis
    1. Note that (1) relays might employ some rate-limiting for block submissions, and (2) blocks are becoming more valuable towards end of the slot, and may be submitted more often then. 
    2. Therefore a smart scheduling might be advisable, sending only few blocks early on in the slot and sending more frequently in the second half of a slot.

### Additional notes

- Geth makes some things very hard to do, and there's some possible strong improvements over geth based on our experience:
    - Fast multi-transaction simulation with cheap reverts, in particular reverting ethereum state to one from before a list of transactions
    - Support for multiple merging algorithms is important, perhaps they can even be made somewhat modular?
    - A pain-point with geth is the slow calculation of the state root and block hash.
- Bundles require revert protection and cannot be unbundled. (bundles are a list of transactions that can either be included together or not at all, and which must not be included if one of the transactions reverts)
- See also the [Flashbots builder README](https://github.com/flashbots/builder) for inspiration on command-line arguments
- The original Flashbots builder implementation uses a custom Prysm fork to send payload attributes to the builder. We are currently in the process of phasing this out and switching to the newly standardized SSE events: [https://github.com/flashbots/builder/pull/53](https://github.com/flashbots/builder/pull/53)
- We are happy to collaborate, feel free reach out through [Discord](https://discord.com/invite/3TjWjBerRb), Telegram or the [Flashbots forum](https://collective.flashbots.net/).

**Comment by onbjerg** [comment]
Duplicate of https://github.com/paradigmxyz/reth/issues/1013, but since this is more fleshed out I will close the other issue

**Comment by gakonst** [comment]
Thank you Chris. Excited to start working on this soon...pinging searchers/builders to weigh in as well.

**Comment by gakonst** [comment]
Makes a ton of sense, thanks @jasalper. Should there also be extra parameters supplied which allow for state overrides?

**Comment by joroshiba** [comment]
@jasalper this would certainly fit our needs, and fits relatively nicely with my suggested interface in #1013 ([here](https://github.com/paradigmxyz/reth/issues/1013#issuecomment-1419487486)).

To implement the Engine API, however, I believe Reth will require the ability to build a block from its mempool? Both v1 and v2 of `eth_forkchoiceUpdate` accept payload attributes and then start the building process. See the spec [here](https://github.com/ethereum/execution-apis/blob/769c53c94c4e487337ad0edea9ee0dce49c79bfa/src/engine/specification.md#engine_forkchoiceupdatedv1), notably point number 8 of the requirements, as well as the details on [payload building](https://github.com/ethereum/execution-apis/blob/769c53c94c4e487337ad0edea9ee0dce49c79bfa/src/engine/specification.md#payload-building).

I think the key here is that as opposed to Geth, we can make Reth easy for people with alternative transaction selection criteria to build their own payloads (and thus be more usable as a library).

**Comment by jasalper** [comment]
@gakonst - I don't think the state overrides would be relevant for us specifically, but I could see `reth_buildBlockWithoutStateRoot` perhaps using them. It doesn't make much sense though to have state overrides but also compute state root/block hash.

**Comment by metachris** [comment]
Just an API call which requires transactions as input and returns a block seems entirely insufficient.

Building blocks for a slot is not an one-off task, but an ongoing effort to find the highest value block by trying (simulating) combinations of transactions _and_ bundles (that aren't allowed to be unbundled), including new ones that arrive over time (from mempool, through `eth_sendBundle`, etc.), and continuously submitting them to the relays.

**Comment by rakita** [comment]
> Duplicate of #1013, but since this is more fleshed out I will close the other issue

Let us leave it open. This issue is more discussion on API and that issue is for implementation and how it gets integrated, so there is a difference.

**Comment by jasalper** [comment]
@metachris - I don't think reth should be making a block builder - just adding support for external builders. I think there is a misunderstanding of what reth does in my proposal. Here, reth does not try to find the most valuable block. reth simply exposes a call that computes the block sealing values for builders that have already computed the most valuable block, and are submitting this block via a list of transactions. 

"Building blocks for a slot is not an one-off task, but an ongoing effort to find the highest value block by trying (simulating) combinations of transactions and bundles (that aren't allowed to be unbundled), including new ones that arrive over time (from mempool, through eth_sendBundle, etc.), and continuously submitting them to the relays."

The RPC call doesn't broadcast the block anywhere. The RPC call is used to compute the state root (+ logs bloom, transactions root, etc) for a given list of transactions. An efficient block builder will not be computing these cryptographic values for any potential block, only the ones they wish to submit. 

In this model, builders would do their simulation via something more efficient like `trace_callMany`, and would continuously update their blocks with incoming transactions outside of reth.  `reth_buildBlock` is only called once a potential block is ready to be sent to the relay. 

"It's also important to point out that bundles need to be supported, which are a list of transactions that can either be included all together or none at all, and which must not be included if one of the transactions reverts."

I don't think `reth_buildBlock` needs to care about whether something is a bundle or a mempool transaction. Once something becomes part of the EL and lands on chain, these distinctions disappear. Ensuring that bundle transactions do not revert and that private transactions remain private should be the responsibility of the block builder. (Although perhaps returning an additional data structure which indicates which transactions reverted might be useful for these purposes?)

"If an external service would need to go through this API call, it would need to be connected to the mempool, and receive bundles/private transactions through additional APIs, and would be very inefficient because of the request/response pattern, even if the API would allow for revert protection, transactions, private transactions and bundles, and even if reth would try to find the most valuable block out of all the supplied inputs while adhering to the expressed preferences."

Having external builders subscribe to the txnpool to pack the block doesn't seem like a particularly big ask - everybody is doing it anyways, and if they don't wish to do it, the `reth_buildBlockWithTxnpool` call would fill the rest of the block with transactions that reth has seen in the default geth-style gas price ordering. (Not optimal compared to other packings, but if a builder wants efficient packings, they should do it themselves). 

The builder already receives bundles/private transactions through additional APIs, so I'm not sure what the distinction is.

**Comment by gakonst** [comment]
I think we can close this given the `CliExt` trait which allows extending things and given we have the `PayloadBuilder` trait now.

also see integrations:

https://github.com/ralexstokes/mev-rs/pull/112/files
https://github.com/jacobkaufmann/evangelion/
